numpy/core/src/multiarray/item_selection.c

```diff
@@ -2096,10 +2103,13 @@ PyArray_CountNonzero(PyArrayObject *self)
             if (nonzero(data, self)) {
                 ++nonzero_count;
             }
+            if (needs_api && PyErr_Occurred()) {
+                return -1;
+            }
             data += stride;
         }
 
-        return nonzero_count;
+        return PyErr_Occurred() ? -1 : nonzero_count;
```

Comment:
What's the rationale for checking in the loop only if an API is needed, but checking outside always?

Are non-`NPY_NEEDS_PYAPI` types allowed to throw errors? If not, then I think this should be more like

```
if(PyErr_Occured()) throw_SystemError("Must set that flag thing");
```